File: samples/feed-demo-ios/SellwildSample/ListingsScreen.swift
import SwiftUI
import UIKit
import SellwildSDK

/// Loads listings with `SellwildAPIClient` for an app that draws its own
/// cards. Refresh clears the client's cache and fetches again.
@MainActor
final class ListingsModel: ObservableObject {
    @Published private(set) var listings: [SellwildListing] = []
    @Published private(set) var photos: [String: UIImage] = [:]
    @Published private(set) var status = "Loading listings"
    private var loads = 0

    func load(config: SellwildConfig, clearCache: Bool) async {
        if clearCache { SellwildAPIClient.shared.clearCache() }
        loads += 1
        let load = loads
        status = "Loading listings (load \(load))"
        do {
            let response = try await SellwildAPIClient.shared.fetchListings(config: config)
            listings = response.listings
            photos = Self.inlinePhotos(response.listings)
            status = "\(response.listings.count) listings, load \(load)"
        } catch {
            // The SDK has already reported this failure; the app only shows it.
            status = "Listings failed (load \(load)): \(error.localizedDescription)"
        }
    }

    /// The feed's photos are data: URIs. Decode them once, not on every render.
    private static func inlinePhotos(_ listings: [SellwildListing]) -> [String: UIImage] {
        var photos: [String: UIImage] = [:]
        for listing in listings {
            guard let url = listing.primaryPhoto?.url, url.hasPrefix("data:"),
                  let comma = url.firstIndex(of: ","),
                  let data = Data(base64Encoded: String(url[url.index(after: comma)...])),
                  let image = UIImage(data: data) else { continue }
            photos[listing.id] = image
        }
        return photos
    }
}

/// Listings: the listings API client and the app's own list of cards.
struct ListingsScreen: View {
    let config: SellwildConfig
    @StateObject private var model = ListingsModel()
    @Environment(\.openURL) private var openURL

    var body: some View {
        VStack(spacing: 0) {
            HStack(alignment: .top) {
                ScreenHeader(title: "Listings", detail: "SellwildAPIClient.fetchListings, drawn by the app.")
                Button("Refresh") {
                    Task { await model.load(config: config, clearCache: true) }
                }
                .buttonStyle(.borderedProminent)
                .padding(.top, 12)
                .padding(.trailing, 16)
                .accessibilityIdentifier(SampleID.listingsRefresh)
            }
            Text(model.status)
                .font(.caption)
                .foregroundColor(.secondary)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.horizontal, 16)
                .padding(.bottom, 8)
                .accessibilityIdentifier(SampleID.listingsStatus)
            List(model.listings) { listing in
                Button {
                    if let link = listing.tapURL(partnerCode: config.partnerCode), let url = URL(string: link) {
                        openURL(url)
                    }
                } label: {
                    ListingCard(listing: listing, photo: model.photos[listing.id])
                }
                .buttonStyle(.plain)
                .accessibilityElement(children: .combine)
                .accessibilityIdentifier(SampleID.listingCard)
            }
            .listStyle(.plain)
            .refreshable { await model.load(config: config, clearCache: true) }
            .accessibilityIdentifier(SampleID.listingsList)
        }
        .task {
            if model.listings.isEmpty { await model.load(config: config, clearCache: false) }
        }
    }
}

/// One listing: photo, title, price and seller.
struct ListingCard: View {
    let listing: SellwildListing
    let photo: UIImage?

    var body: some View {
        HStack(spacing: 12) {
            Group {
                if let photo {
                    Image(uiImage: photo).resizable().scaledToFill()
                } else if let link = listing.primaryPhoto?.url, link.hasPrefix("http"), let url = URL(string: link) {
                    AsyncImage(url: url) { image in
                        image.resizable().scaledToFill()
                    } placeholder: {
                        Color.gray.opacity(0.2)
                    }
                } else {
                    Color.gray.opacity(0.2)
                }
            }
            .frame(width: 72, height: 72)
            .clipShape(RoundedRectangle(cornerRadius: 8))

            VStack(alignment: .leading, spacing: 4) {
                Text(listing.title).font(.subheadline).lineLimit(2)
                if let price = listing.displayPrice {
                    Text("$\(price)").font(.footnote.weight(.semibold))
                }
                if let seller {
                    Text("by \(seller)").font(.caption).foregroundColor(.secondary)
                }
            }
            Spacer(minLength: 0)
        }
        .padding(.vertical, 4)
    }

    /// "Nick P.", from the seller's first name and last initial.
    private var seller: String? {
        let first = listing.user?.firstName?.trimmingCharacters(in: .whitespaces) ?? ""
        guard !first.isEmpty else { return nil }
        guard let initial = listing.user?.lastName?.first else { return first }
        return "\(first) \(initial)."
    }
}

File: samples/flutter-demo/lib/src/listings_screen.dart
import 'dart:async';
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'package:sellwild_sdk/sellwild_sdk.dart';

import 'sample_ids.dart';
import 'widgets.dart';

/// The listings status after a good load (contracts/e2e/ids.json,
/// sw.listings.status): '<count> listings, load <n>'.
String listingsStatus(int count, int load) => '$count listings, load $load';

/// The bytes of a `data:` photo URL, or null when [url] is no data URL or
/// does not decode. The feed's photos are data URLs.
Uint8List? dataUrlBytes(String? url) {
  if (url == null || !url.startsWith('data:')) return null;
  try {
    return UriData.parse(url).contentAsBytes();
  } on FormatException {
    return null;
  }
}

/// Listings: SellwildAPIClient.fetchListings, drawn by the app as its own
/// list of cards. Refresh clears the client's cache and fetches again.
class ListingsScreen extends StatefulWidget {
  const ListingsScreen({super.key, required this.config});

  final SellwildConfig config;

  @override
  State<ListingsScreen> createState() => _ListingsScreenState();
}

class _ListingsScreenState extends State<ListingsScreen> {
  List<SellwildListing> _listings = const [];
  Map<String, Uint8List> _photos = const {};
  String _status = 'Loading listings';
  int _loads = 0;

  @override
  void initState() {
    super.initState();
    unawaited(_load(clearCache: false));
  }

  Future<void> _load({required bool clearCache}) async {
    final client = SellwildAPIClient.instance;
    if (clearCache) client.clearCache();
    final load = ++_loads;
    setState(() => _status = 'Loading listings (load $load)');
    try {
      final response = await client.fetchListings(widget.config);
      if (!mounted) return;
      // Decode the data URL photos once, not on every build.
      final photos = <String, Uint8List>{
        for (final listing in response.listings)
          if (dataUrlBytes(listing.primaryPhoto?.url) case final bytes?)
            listing.id: bytes,
      };
      setState(() {
        _listings = response.listings;
        _photos = photos;
        _status = listingsStatus(response.listings.length, load);
      });
    } catch (e) {
      // The SDK has already reported this failure; the app only shows it.
      if (!mounted) return;
      setState(() => _status = 'Listings failed (load $load): $e');
    }
  }

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      children: [
        Row(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            const Expanded(
              child: ScreenHeader(
                title: 'Listings',
                detail: 'SellwildAPIClient.fetchListings, drawn by the app.',
              ),
            ),
            Padding(
              padding: const EdgeInsets.only(top: 12, right: 16),
              child: E2EId(
                SampleId.listingsRefresh,
                child: FilledButton(
                  onPressed: () => unawaited(_load(clearCache: true)),
                  child: const Text('Refresh'),
                ),
              ),
            ),
          ],
        ),
        StatusLine(SampleId.listingsStatus, _status),
        Expanded(
          child: E2EId(
            SampleId.listingsList,
            label: 'Listings',
            child: RefreshIndicator(
              onRefresh: () => _load(clearCache: true),
              child: ListView.separated(
                itemCount: _listings.length,
                separatorBuilder: (_, __) => const Divider(height: 1),
                itemBuilder: (context, i) {
                  final listing = _listings[i];
                  return E2EId(
                    SampleId.listingCard,
                    child: _ListingRow(
                      listing: listing,
                      photo: _photos[listing.id],
                      onTap: () => unawaited(openLink(context, listing.url)),
                    ),
                  );
                },
              ),
            ),
          ),
        ),
      ],
    );
  }
}

/// One listing: photo, title, price and seller.
class _ListingRow extends StatelessWidget {
  const _ListingRow({
    required this.listing,
    required this.photo,
    required this.onTap,
  });

  final SellwildListing listing;
  final Uint8List? photo;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final price = listing.displayPrice;
    final seller = sellerName(listing.user);
    return InkWell(
      onTap: onTap,
      child: Padding(
        padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 8),
        child: Row(
          children: [
            ClipRRect(
              borderRadius: BorderRadius.circular(8),
              child: SizedBox(width: 72, height: 72, child: _photo(theme)),
            ),
            const SizedBox(width: 12),
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(
                    listing.title,
                    maxLines: 2,
                    overflow: TextOverflow.ellipsis,
                    style: theme.textTheme.bodyMedium,
                  ),
                  if (price != null)
                    Text('\$$price', style: theme.textTheme.labelLarge),
                  if (seller != null)
                    Text('by $seller', style: theme.textTheme.bodySmall),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _photo(ThemeData theme) {
    final placeholder =
        ColoredBox(color: theme.colorScheme.surfaceContainerHighest);
    final bytes = photo;
    if (bytes != null) {
      // A format the platform cannot decode (AVIF on some) shows the
      // placeholder.
      return Image.memory(
        bytes,
        fit: BoxFit.cover,
        errorBuilder: (_, __, ___) => placeholder,
      );
    }
    final link = listing.primaryPhoto?.url;
    if (link != null && link.startsWith('http')) {
      return Image.network(
        link,
        fit: BoxFit.cover,
        errorBuilder: (_, __, ___) => placeholder,
      );
    }
    return placeholder;
  }
}

/// "Nick P.": the seller's first name and last initial, or null.
String? sellerName(SellwildUser? user) {
  final first = user?.firstName.trim() ?? '';
  if (first.isEmpty) return null;
  final last = user?.lastName.trim() ?? '';
  return last.isEmpty ? first : '$first ${last[0]}.';
}
